plugin/evm/vm.go

```diff
@@ -1271,6 +1279,50 @@ func (vm *VM) initBlockBuilding() error {
 	return nil
 }
 
+func (vm *VM) SubscribeToEvents(ctx context.Context) commonEng.Message {
+	// Block building is not initialized yet, so we haven't finished syncing or bootstrapping.
+	if vm.builder == nil {
+		select {
+		case <-ctx.Done():
+			return 0
+		case ss := <-vm.stateSyncDone:
+			return ss
+		case <-vm.shutdownChan:
+			return 0
+		case <-vm.finishedSyncing:
+			break
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pending := make(chan commonEng.Message, 1)
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		pending <- vm.builder.waitForTxEnqueue(ctx)
+	}()
+
+	defer wg.Wait()
+
+	select {
+	case ss := <-vm.stateSyncDone:
+		return ss
```

Comment:
What guarantees `wg.Wait()` to return when this case is hit? It seems to me that this could cause a deadlock.

Reply:
You're right, I should've had the `wg.Wait()` and the `cancel()` in the opposite order. Fixed, thanks. 

However when I wrote the code I concluded that this case can never happen, as if `vm.builder` is not nil, then we finished bootstrapping, and if we finished bootstrapping then this method had read the notification in the select case above where `vm.builder` is nil. I still left this case for the sake of completeness.

Reply:
I think there is still an issue here because `waitForTxEnqueue` may be blocking on `pendingSignal` (and so the cancellation doesn't do anything)

Reply:
oh right, good catch. [Fixed](https://github.com/ava-labs/coreth/pull/1003/commits/88d4001d06f2fd7cb89ebbae37db1bc6392d98f0).